src/xb-node-query.c

```diff
@@ -87,15 +121,51 @@ xb_node_query_full (XbNode *self, XbQuery *query, GError **error)
  **/
 XbNode *
 xb_node_query_first_full (XbNode *self, XbQuery *query, GError **error)
+{
+	return xb_node_query_first_with_context (self, query, NULL, error);
+}
+
+/**
+ * xb_node_query_first_with_context:
+ * @self: a #XbNode
+ * @query: an #XbQuery
+ * @context: (nullable) (transfer none): context including values bound to opcodes of type
+ *     %XB_OPCODE_KIND_BOUND_INTEGER or %XB_OPCODE_KIND_BOUND_TEXT, or %NULL if
+ *     the query doesn’t need any context
+ * @error: the #GError, or %NULL
+ *
+ * Searches the silo using a prepared query, returning up to one result.
+ *
+ * It is safe to call this function from a different thread to the one that
+ * created the #XbSilo.
+ *
+ * Please note: Only a subset of XPath is supported.
+ *
+ * Returns: (transfer full): a #XbNode, or %NULL if unfound
+ *
+ * Since: 0.3.0
+ **/
+XbNode *
+xb_node_query_first_with_context (XbNode *self, XbQuery *query, XbQueryContext *context, GError **error)
 {
 	g_autoptr(GPtrArray) results = NULL;
+	guint old_limit;
 
 	g_return_val_if_fail (XB_IS_NODE (self), NULL);
 	g_return_val_if_fail (XB_IS_QUERY (query), NULL);
 	g_return_val_if_fail (error == NULL || *error == NULL, NULL);
 
+	if (context != NULL) {
+		old_limit = xb_query_context_get_limit (context);
+		xb_query_context_set_limit (context, 1);
+	}
+
 	/* nodes don't have to include themselves as part of the query */
-	results = xb_silo_query_with_root_full (xb_node_get_silo (self), self, query, error);
+	results = xb_silo_query_with_root_full (xb_node_get_silo (self), self, query, context, error);
+
+	if (context != NULL)
+		xb_query_context_set_limit (context, old_limit);
```

Comment:
this smells fishy, no?

Reply:
It does. I was struggling to find a way to make it better. I’ve pushed a commit which might improve things — see what you think.